--- gemmini.h
+++ gemmini.h
@@ -1102,10 +1102,18 @@
         acc_t sum_exp = 0;
         for (size_t j = 0; j < DIM_J; j++) {
           acc_t q = c_buffer[j] - max_q;
+#ifdef ELEM_T_IS_FLOAT
+          acc_t z = (acc_t) (-q * qln2_inv) / 65536.0f;
+#else
           acc_t z = (acc_t) (-q * qln2_inv) >> 16;
+#endif
           acc_t qp = q + z * qln2;
           acc_t q_exp = (qp + qb)*(qp + qb) + qc;
+#ifdef ELEM_T_IS_FLOAT
+          c_buffer[j] = (acc_t) ldexp((double) q_exp, -(int) z);
+#else
           c_buffer[j] = q_exp >> z;
+#endif
           sum_exp += c_buffer[j];
         }
 
